### ISSUE TITLE
Uploading the same image for multiple parts results in multiple images being safed

### ISSUE THREAD

**Issue by 30350n** [inventree/InvenTree#5532]
### Please verify that this bug has NOT been raised before.

- [X] I checked and didn't find a similar issue

### Describe the bug*

I've added a whole bunch of parts to my instance (via an automatic script that fetches the parts from digikey) and noticed a lot of media files accumulating. Thus I wondered if uploading the same image twice would result in two media files being saved.

I just did a quick check by running a python script over all part_images, hashing their content and confirmed that out of the 360 files I've got, only 248 are unique.

### Steps to Reproduce

1. upload the same image for two different parts
2. observe to files being created in the media/part_images directory

### Expected behaviour

Files should only be stored if they contain unique data.

### Deployment Method

- [ ] Docker
- [X] Bare metal

### Version Information

# Version Information:
InvenTree-Version: 0.13.0 dev
Django Version: 3.2.20
Commit Hash: cd35744
Commit Date: 2023-09-12

Database: postgresql
Debug-Mode: False
Deployed using Docker: False
Platform: Linux-3.10.0-1160.83.1.el7.x86_64-x86_64-with-glibc2.17
Installer: GIT

Active plugins: [{'name': 'InvenTreeBarcode', 'slug': 'inventreebarcode', 'version': '2.0.0'}, {'name': 'InvenTreeCoreNotificationsPlugin', 'slug': 'inventreecorenotificationsplugin', 'version': '1.0.0'}, {'name': 'InvenTreeLabel', 'slug': 'inventreelabel', 'version': '1.0.0'}, {'name': 'DigiKeyPlugin', 'slug': 'digikeyplugin', 'version': '1.0.0'}, {'name': 'LCSCPlugin', 'slug': 'lcscplugin', 'version': '1.0.0'}, {'name': 'MouserPlugin', 'slug': 'mouserplugin', 'version': '1.0.0'}]

### Please verify if you can reproduce this bug on the demo site.

- [ ] I can reproduce this bug on the demo site.

**Comment by 30350n** [comment]
Opening this issue because I'm not sure if there's a solid reason for this behavior.

I can see two possible solutions to fix this:
1. just hash every image after uploading and save it with the hash as its name (simplest, but doesn't allow for keeping the file name)
2. basically 1. but keep some kind of dictionary to map from image name to hash

If this would be desirable behavior, I'd be happy to look into implementing this :)

**Comment by SchrodingersGat** [comment]
Currently the way to have multiple parts use the same image is to use the "select existing image" process - have you been using this? 

1. Create a part A
2. Upload an image for the part
3. Create a part B
4. Select the existing image (from part A) to use for part B

Hashing each image and checking *all* existing images would be potentially:

a) tricky
b) time consuming
c) Introduce extra failure modes / edge cases

Perhaps you can hook into the existing "select existing image" functionality to achieve what you want here?

**Comment by 30350n** [comment]
> Perhaps you can hook into the existing "select existing image" functionality to achieve what you want here?

Well there's two problems with that.
1. I'm adding the images automatically via a script, so that's just not possible I think.
2. This is how the Select Part Image dialog looks like for me, for whatever reason:
![image](https://github.com/inventree/InvenTree/assets/34186858/efe00c3b-e75f-49ee-874f-872dce5980a0)

What would be tricky/time consuming about hashing the images exactly? Hashing the 360 part images I got took less than a second, so I kinda doubt that be an issue? Edge cases are what I'd probably be most concerned about, but I can't think of any really. (Except for the obvious issue that the image names would be messed up, but the way it is now they already get changed so yeah).

**Comment by SchrodingersGat** [comment]
> What would be tricky/time consuming about hashing the images exactly?

We would need to create new database schema for storing the image hashes and then comparing on upload. It seems like a lot of complication for (IMO) minimal benefit? You might be able to convince me elsewise.

I'm hoping there is a simpler way :)

**Comment by simonkuehling** [comment]
maybe a `ScheduleMixin` plugin could do the job here, periodically running a cleanup-task to consolidate duplicates of images?

@SchrodingersGat I can see references to `django_cleanup` in the code, which could lead to unexpected results since this does not check for multiple models referencing the same media file if I understood the docs correctly...?

**Comment by SchrodingersGat** [comment]
@simonkuehling we use the `@cleanup.ignore` decorator to protect part and company models from being incorrectly "cleaned" by the `django_cleanup` librayr

**Comment by simonkuehling** [comment]
ah, ok. So, if a plugin would modify the image url of some parts to eliminate duplicates - that would not automatically do anything to the media files folder, right? (The duplicate images would still be in the media dir)

(only of interest if such a plugin solution would be suitable for @30350n in this case, of course)

**Comment by 30350n** [comment]
Could probably work, but I still think just naming the images by image hash would be the better and easier to implement solution. The image names already get alterered anyways, all you'd have to do really is just replace the whole file name with an actual file hash.
![image](https://github.com/inventree/InvenTree/assets/34186858/cea74612-1cce-48d6-bc5d-937ae1ecac5f)
I just hadn't time to look at the exact implemenation yet.

**Comment by wolflu05** [comment]
In my opinion the best and scalable solution would be to name them by a uuid and create a db table images which is polymorphic related to parts, notes images, ... (everywhere where images are used) and then instead of messing with the filesystem on our self, we use the Django solution which also supports other storage types like s3, then there could also be implemented direct serving if the storage provider supports it. Same could be done for other attachments like files.

This would provably also fix #3029, #2585

**Comment by 30350n** [comment]
> The thing with hashes is also, that there are 3 types of images stored for different previewing sizes.

That is not an issue though.

**Comment by SchrodingersGat** [comment]
We could potentially rework the existing upload function to use the file content hash as the image name, and then check if an image with that filename exists. 

(We should then also double check that the hash does actually match).

In such a case, we can simply point to that image, rather than saving the duplicate.

**Comment by 30350n** [comment]
Yep, that is pretty much exactly what I wanted to do. I just hadn't had much time to look into this yet and with the little time I spent, I honestly have struggled to even find the function responsible for the renaming/writing 😅 .
If you could guide me to it, I can probably throw together a quick PR the next few days ^^

**Comment by 30350n** [comment]
Also, I wonder if you can do something like a migration on media files or if that's not possible/wouldn't be a good idea.

**Comment by SchrodingersGat** [comment]
I would not suggest touching any existing files - only run this "hash based naming" scheme for newly uploaded files.

**Comment by 30350n** [comment]
Not stale, being worked on in #5703.

**Comment by SchrodingersGat** [comment]
We could add a write only field to the API which allows an existing part image path to be specified. If the specific image does not exist, or is outside the part image directory it throws an error.

**Comment by 30350n** [comment]
Issue here would be obtaining said image reference in the first place though. At least for my use case, I'd have to search the image "by data" still (which could more or less be done by naming the image after it's hash on the client side but yeah).

**Comment by SchrodingersGat** [comment]
@30350n see https://github.com/inventree/InvenTree/pull/6021 which implements this feature

**Comment by asnreadily3d** [comment]
# Docker compose recipe for a production-ready InvenTree setup, with the following containers:
# - PostgreSQL as the database backend
# - gunicorn as the InvenTree web server
# - django-q as the InvenTree background worker process
# - Caddy as a reverse proxy
# - redis as the cache manager (optional, disabled by default)

# ---------------------
# READ BEFORE STARTING!
# ---------------------

# -----------------------------
# Setting environment variables
# -----------------------------
# Shared environment variables should be stored in the .env file
# Changes made to this file are reflected across all containers!
#
# IMPORTANT NOTE:
# You should not have to change *anything* within this docker-compose.yml file!
# Instead, make any changes in the .env file!

# ------------------------
# InvenTree Image Versions
# ------------------------
# By default, this docker-compose script targets the STABLE version of InvenTree,
# image: inventree/inventree:stable
#
# To run the LATEST (development) version of InvenTree,
# change the INVENTREE_TAG variable (in the .env file) to "latest"
#
# Alternatively, you could target a specific tagged release version with (for example):
# INVENTREE_TAG=0.7.5
#

# ----------------------------
# Docker compose customization
# ----------------------------
# If you wish to customize the docker-compose script, you should only do so if you understand the stack!
# Do not expect support for customizations that are not part of the standard InvenTree setup!

services:
    # Database service
    # Use PostgreSQL as the database backend
    inventree-db:
        image: postgres:16
        container_name: inventree-db
        expose:
            - ${INVENTREE_DB_PORT:-5432}/tcp
        environment:
            - PGDATA=/var/lib/postgresql/data/pgdb
            - POSTGRES_USER=${INVENTREE_DB_USER:?You must provide the 'INVENTREE_DB_USER' variable in the .env file}
            - POSTGRES_PASSWORD=${INVENTREE_DB_PASSWORD:?You must provide the 'INVENTREE_DB_PASSWORD' variable in the .env file}
            - POSTGRES_DB=${INVENTREE_DB_NAME:?You must provide the 'INVENTREE_DB_NAME' variable in the .env file}
        volumes:
            # Map 'data' volume such that postgres database is stored externally
            - ${INVENTREE_EXT_VOLUME:?You must specify the 'INVENTREE_EXT_VOLUME' variable in the .env file!}:/var/lib/postgresql/data/:z
        restart: unless-stopped

    # redis acts as database cache manager
    inventree-cache:
        image: redis:7-alpine
        container_name: inventree-cache
        env_file:
            - .env
        expose:
            - ${INVENTREE_CACHE_PORT:-6379}
        restart: always

    # InvenTree web server service
    # Uses gunicorn as the web server
    inventree-server:
        # If you wish to specify a particular InvenTree version, do so here
        image: inventree/inventree:${INVENTREE_TAG:-stable}
        container_name: inventree-server
        # Only change this port if you understand the stack.
        expose:
            - 8000
        depends_on:
            - inventree-db
            - inventree-cache
        env_file:
            - .env
        volumes:
            # Data volume must map to /home/inventree/data
            - ${INVENTREE_EXT_VOLUME}:/home/inventree/data:z
        restart: unless-stopped

    # Background worker process handles long-running or periodic tasks
    inventree-worker:
        # If you wish to specify a particular InvenTree version, do so here
        image: inventree/inventree:${INVENTREE_TAG:-stable}
        container_name: inventree-worker
        command: invoke worker
        depends_on:
            - inventree-server
        env_file:
            - .env
        volumes:
            # Data volume must map to /home/inventree/data
            - ${INVENTREE_EXT_VOLUME}:/home/inventree/data:z
        restart: unless-stopped

    # caddy acts as reverse proxy and static file server
    # https://hub.docker.com/_/caddy
    inventree-proxy:
        container_name: inventree-proxy
        image: caddy:alpine
        restart: always
        depends_on:
            - inventree-server
        ports:
            - ${INVENTREE_WEB_PORT:-80}:80
            - 443:443
        env_file:
            - .env
        volumes:
            - ./Caddyfile:/etc/caddy/Caddyfile:ro,z
            - ${INVENTREE_EXT_VOLUME}/static:/var/www/static:z
            - ${INVENTREE_EXT_VOLUME}/media:/var/www/media:z
            - ${INVENTREE_EXT_VOLUME}:/var/log:z
            - ${INVENTREE_EXT_VOLUME}:/data:z
            - ${INVENTREE_EXT_VOLUME}:/config:z
